sched/queue/memory/simple.go

```diff
@@ -3,40 +3,134 @@ package memory
 // In-memory Scheduler Queue
 
 import (
-	"fmt"
 	"github.com/scootdev/scoot/sched"
 	"github.com/scootdev/scoot/sched/queue"
+	"strconv"
 	"time"
 )
 
-func NewSimpleQueue() (queue.Queue, chan queue.WorkItem) {
+func NewSimpleQueue(capacity int) queue.Queue {
 	q := &simpleQueue{}
-	q.outCh = make(chan queue.WorkItem, 1)
-	q.inCh = make(chan enqueueReq)
-	go q.loop()
-	return q, q.outCh
+	q.reqCh = make(chan queueReq)
+	q.outCh = make(chan queue.WorkItem)
+
+	st := &simpleQueueState{reqCh: q.reqCh, outCh: q.outCh, capacity: capacity}
+	go st.loop()
+	return q
 }
 
 type simpleQueue struct {
-	nextID int
-	inCh   chan enqueueReq
-	outCh  chan queue.WorkItem
+	// Immutable state, read by many goroutines
+	reqCh chan queueReq
+	outCh chan queue.WorkItem
+}
+
+type simpleQueueState struct {
+	// Mutable state, only read/written by its goroutine
+	reqCh     chan queueReq
+	outCh     chan queue.WorkItem
+	dequeueCh chan struct{}
+	items     []*simpleWorkItem
+	nextID    int
+	capacity  int
+}
+
+func (s *simpleQueueState) loop() {
+	for !s.done() {
+		s.iter()
+
+	}
+	close(s.outCh)
+}
+
+func (s *simpleQueueState) done() bool {
+	// We are done once our inputs are closed and we have no items left to send
+	return s.reqCh == nil && len(s.items) == 0
 }
 
-type simpleWorkItem sched.Job
+func (s *simpleQueueState) iter() {
```

Comment:
Wait, this has back-pressure, right? Because it checks it against capacity? Or am I misunderstanding?

As for possibly-nil outCh: I found Dave Cheney's explanation of "Channel Axioms" very useful: http://dave.cheney.net/2014/03/19/channel-axioms

In this case, "A send to a nil channel blocks forever"

So, sometimes we don't want to send to outCh (if we have nothing to send, or if we're still waiting for a dequeue ping to come in). How do we do this?

One way: use if to pick which of several select statements to call. BUT, then the select statements have a lot of repeated code. So that's bad.

We could use reflect.Select, which takes a slice of SelectCase. But that leads to code that looks weird.

So, AIUI (and maybe I'm missing a simpler way), if there's some case in a select statement you want to do conditionally, the idiomatic way is to set it to nil (in which case the operation will block forever, and therefore never be selected).